File: Java/app/src/main/java/software/rsquared/template/utils/AutoClearedValue.java
package software.rsquared.template.utils;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;

/**
 * @author rSquared.software
 */

public class AutoClearedValue<T> {

    private OnClearListener<T> clearListener;

    private T value;

    public AutoClearedValue(Fragment fragment, T value) {
        FragmentManager fragmentManager = fragment.getFragmentManager();
        if (fragmentManager != null) {
            fragmentManager.registerFragmentLifecycleCallbacks(
                    new FragmentManager.FragmentLifecycleCallbacks() {
                        @Override
                        public void onFragmentViewDestroyed(FragmentManager fm, Fragment f) {
                            if (f == fragment) {
                                destroy();
                                fragmentManager.unregisterFragmentLifecycleCallbacks(this);
                            }
                        }
                    }, false);
            this.value = value;
        }
    }

    private void destroy() {
        if (clearListener != null) {
            clearListener.onClear(this.value);
        }
        this.value = null;
    }

    public void setOnClearListener(OnClearListener<T> clearListener) {
        this.clearListener = clearListener;
    }

    public T get() {
        return value;
    }

    public interface OnClearListener<T> {
        void onClear(T value);
    }
}

File: app/src/main/java/software/rsquared/template/utils/AutoClearedValue.kt
package software.rsquared.template.utils

import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager

class AutoClearedValue<T>(fragment: Fragment, var value: T?) {

    init {
        fragment.fragmentManager?.registerFragmentLifecycleCallbacks(
                object : FragmentManager.FragmentLifecycleCallbacks() {
                    override fun onFragmentViewDestroyed(fm: FragmentManager?, f: Fragment?) {
                        if (f === fragment) {
                            value = null
                            fragment.fragmentManager?.unregisterFragmentLifecycleCallbacks(this)
                        }
                    }
                }, false)
    }
}